Tests/DatadogTests/Datadog/URLSessionAutoInstrumentation/Interception/AppStateListenerTests.swift

```diff
@@ -0,0 +1,201 @@
+            finalState: .init(isInBackground: false, date: startDate.addingTimeInterval(10.0))
+        )
+        XCTAssertEqual(limitedHistory, expectedHistory)
+    }
+}
+
+class NonUIKitAppStateListenerTests: XCTestCase {
+    func testListener() {
+        let listener = NonUIKitAppStateListener()
+
+        let history1 = listener.appStateHistory
+        XCTAssert(history1.changes.isEmpty)
+
+        let history2 = listener.appStateHistory
+        XCTAssertEqual(history2.initialState, history1.initialState)
+        XCTAssertGreaterThan(history2.finalState.date, history1.finalState.date)
+
+        XCTAssertFalse(history1.initialState.isInBackground)
+        XCTAssertFalse(history1.finalState.isInBackground)
+        XCTAssertFalse(history2.initialState.isInBackground)
+        XCTAssertFalse(history2.finalState.isInBackground)
+    }
+}
+
+#if canImport(UIKit)
+class AppStateListenerTests: XCTestCase {
+    func testUIKitAppListener() {
+        let startDate = Date.mockDecember15th2019At10AMUTC()
+        let listener = AppStateListener(
+            dateProvider: RelativeDateProvider(startingFrom: startDate, advancingBySeconds: 1.0)
+        )
+
+        NotificationCenter.default.post(name: UIApplication.didEnterBackgroundNotification, object: nil)
+        NotificationCenter.default.post(name: UIApplication.willEnterForegroundNotification, object: nil)
+        let history = listener.appStateHistory
+
+        let expected = AppStateHistory(
+            initialState: .init(isInBackground: false, date: startDate),
+            changes: [
+                .init(isInBackground: true, date: startDate.addingTimeInterval(1.0)),
+                .init(isInBackground: false, date: startDate.addingTimeInterval(2.0))
+            ],
+            finalState: .init(isInBackground: false, date: startDate.addingTimeInterval(3.0))
+        )
+        XCTAssertEqual(history, expected)
+    }
+
+    func testUIKitAppListenerWithInverseOrder() {
+        let startDate = Date.mockDecember15th2019At10AMUTC()
+        let listener = AppStateListener(
+            dateProvider: RelativeDateProvider(startingFrom: startDate, advancingBySeconds: 1.0)
+        )
+        NotificationCenter.default.post(name: UIApplication.willEnterForegroundNotification, object: nil)
+        NotificationCenter.default.post(name: UIApplication.didEnterBackgroundNotification, object: nil)
+        let history = listener.appStateHistory
+
+        let expected = AppStateHistory(
+            initialState: .init(isInBackground: false, date: startDate),
+            changes: [
+                .init(isInBackground: false, date: startDate.addingTimeInterval(1.0)),
+                .init(isInBackground: true, date: startDate.addingTimeInterval(2.0))
+            ],
+            finalState: .init(isInBackground: false, date: startDate.addingTimeInterval(3.0))
+        )
+        XCTAssertEqual(history, expected)
+    }
+
+    func testChangingFinalState() {
```

Comment:
The `AppStateListener` has its lifecycle and behaviours, let's use `Given` / `When` / `Then` or `When` / `Then` verbosity for this tests. It's hard to understand what "changing final state" means.